src/platform/backends/libvirt/libvirt_virtual_machine.cpp

```diff
@@ -199,22 +199,33 @@ auto domain_by_definition_for(const mp::VirtualMachineDescription& desc, const s
     return domain;
 }
 
-auto domain_state_for(virDomainPtr domain)
+auto domain_state_for(const mp::VirtualMachine::State& current_instance_state, virDomainPtr domain)
 {
-    auto state{0};
+    auto domain_state{0};
 
-    if (!domain)
+    if (!domain || virDomainGetState(domain, &domain_state, nullptr, 0) == -1 || domain_state == VIR_DOMAIN_NOSTATE)
         return mp::VirtualMachine::State::unknown;
 
     if (virDomainHasManagedSaveImage(domain, 0) == 1)
         return mp::VirtualMachine::State::suspended;
 
-    if (virDomainGetState(domain, &state, nullptr, 0) == -1)
-    {
-        return mp::VirtualMachine::State::unknown;
-    }
+    if (domain_state == VIR_DOMAIN_SHUTOFF)
+        return mp::VirtualMachine::State::off;
+
+    if (domain_state == VIR_DOMAIN_RUNNING && current_instance_state == mp::VirtualMachine::State::off)
+        return mp::VirtualMachine::State::running;
+
+    return current_instance_state;
```

Comment:
Why are you passing Multipass' thinking of the current state through? Can we not read all possible states directly from libvirt?

Reply:
There are a number of Multipass states that have no analogous `libvirt` states such as `DELAYED_SHUTDOWN`, `STARTING`, `SUSPENDING`, and Multipass's way of handling `UNKNOWN`. So in this particular piece of code, this is the "fallthrough" in which we just return what Multipass has stored since no libvirt state matches.

Reply:
I see. Could I request a function name change of `domain_state_for` to something like `resolve_domain_state_with_current` ?

Reply:
I kept thinking the function name morphed into a weak name, and as we all know, naming functions (and variables) is one of the hardest software engineering principles :grinning:, I was drawing a blank on what to rename it.  I'll change it for sure.

Reply:
Ok, I renamed it.  Let me know what you think.